pkg/apis/kubeone/v1beta1/types.go

```diff
@@ -0,0 +1,331 @@
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KubeOneCluster is KubeOne Cluster API Schema
+type KubeOneCluster struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Name is the name of the cluster
+	Name string `json:"name"`
+	// ControlPlane describes the control plane nodes and how to access them
+	ControlPlane ControlPlaneConfig `json:"controlPlane,omitempty"`
+	// APIEndpoint are pairs of address and port used to communicate with the Kubernetes API
+	APIEndpoint APIEndpoint `json:"apiEndpoint,omitempty"`
+	// CloudProvider configures the cloud provider specific features
+	CloudProvider CloudProviderSpec `json:"cloudProvider,omitempty"`
+	// Versions defines which Kubernetes version will be installed
+	Versions VersionConfig `json:"versions,omitempty"`
+	// ClusterNetwork configures the in-cluster networking
+	ClusterNetwork ClusterNetworkConfig `json:"clusterNetwork,omitempty"`
+	// Proxy configures proxy used while installing Kubernetes and by the Docker daemon
+	Proxy ProxyConfig `json:"proxy,omitempty"`
+	// StaticWorkers describes the worker nodes that are managed by KubeOne/kubeadm
+	StaticWorkers StaticWorkersConfig `json:"staticWorkers,omitempty"`
+	// DynamicWorkers descirbes the worker nodes that are managed by
+	// Kubermatic machine-controller/Cluster-API
+	DynamicWorkers []DynamicWorkerConfig `json:"dynamicWorkers,omitempty"`
+	// MachineController configures the Kubermatic machine-controller component
+	MachineController *MachineControllerConfig `json:"machineController,omitempty"`
+	// Features enables and configures additional cluster features
+	Features Features `json:"features,omitempty"`
+	// Addons are used to deploy additional manifests
+	Addons *Addons `json:"addons,omitempty"`
+	// SystemPackages configure kubeone behaviour regarding OS packages
+	SystemPackages *SystemPackages `json:"systemPackages,omitempty"`
+}
+
+// OperatingSystemName defines the operating system used on instances
+type OperatingSystemName string
+
+var (
+	OperatingSystemNameUbuntu  OperatingSystemName = "ubuntu"
+	OperatingSystemNameCentOS  OperatingSystemName = "centos"
+	OperatingSystemNameCoreOS  OperatingSystemName = "coreos"
+	OperatingSystemNameFlatcar OperatingSystemName = "flatcar"
+	OperatingSystemNameUnknown OperatingSystemName = ""
+)
+
+// HostConfig describes a single control plane node.
+type HostConfig struct {
+	ID                int    `json:"-"`
+	PublicAddress     string `json:"publicAddress"`
+	PrivateAddress    string `json:"privateAddress"`
+	SSHPort           int    `json:"sshPort"`
+	SSHUsername       string `json:"sshUsername"`
+	SSHPrivateKeyFile string `json:"sshPrivateKeyFile"`
+	SSHAgentSocket    string `json:"sshAgentSocket"`
+	Bastion           string `json:"bastion"`
+	BastionPort       int    `json:"bastionPort"`
+	BastionUser       string `json:"bastionUser"`
+	Hostname          string `json:"hostname"`
+	IsLeader          bool   `json:"isLeader"`
```

Comment:
@kron4eg WDYT about dropping this field and switching to automatically determining the leader? I think that's important for the `apply` use case.

Reply:
if we can do it, that'd be awesome

Reply:
I've thought about this a bit more. I don't think we can get rid of it because we still need the leader concept in order to easily provision and upgrade the cluster.

We can change the defaulting instead. Instead of choosing the first host as the leader, we would do it at the runtime. We would run a set of commands to determine is there Kubernetes running on the instance and once we find the first instance with Kubernetes running, pick it as the leader. If there is no such instance, pick the first one as the leader.

That would result in the same result as now, but there is one important case handled. If the first instance is not responsible, we would pick the second one instead, and we would be able to do the stuff as usual.

I would consider making the field inaccessible to the users because if some user sets the field to the wrong host, apply might give wrong results. On the other side, if we incorrectly determine the leader, bad things would happen.

What do you think about this? What should we do?

Reply:
Choosing a leader should be done really carefully.
There is a situation when first instance in responsive, but empty. It happens when 1 instances has been rotated/wiped and needs "join back to the cluster" actions.

Reply:
Let's leave this discussion for when we start working on the apply.